README.md
<div>
    <a href="https://travis-ci.org/pjfl/p5-dist-zilla-plugin-test-reportmetadata"><img src="https://travis-ci.org/pjfl/p5-dist-zilla-plugin-test-reportmetadata.svg?branch=master" alt="Travis CI Badge"></a>
    <a href="http://badge.fury.io/pl/Dist-Zilla-Plugin-Test-ReportMetadata"><img src="https://badge.fury.io/pl/Dist-Zilla-Plugin-Test-ReportMetadata.svg" alt="CPAN Badge"></a>
    <a href="http://cpants.cpanauthors.org/dist/Dist-Zilla-Plugin-Test-ReportMetadata"><img src="http://cpants.cpanauthors.org/dist/Dist-Zilla-Plugin-Test-ReportMetadata.png" alt="Kwalitee Badge"></a>
</div>

# Name

Dist::Zilla::Plugin::Test::ReportMetadata - Report on prerequisite versions during automated testing

# Synopsis

    # In dist.ini
    [Test::ReportMetadata]

# Description

The is a clone of [Dist::Zilla::Plugin::Test::ReportPrereqs](https://metacpan.org/pod/Dist::Zilla::Plugin::Test::ReportPrereqs) but with the
dependency on [ExtUtils::MakeMaker](https://metacpan.org/pod/ExtUtils::MakeMaker) replaced with one on
[Module::Metadata](https://metacpan.org/pod/Module::Metadata). If you are using [Module::Build](https://metacpan.org/pod/Module::Build) then [Module::Metadata](https://metacpan.org/pod/Module::Metadata)
is already a dependency

Versions are reported based on the result of the `version` attribute from
[Module::Metadata](https://metacpan.org/pod/Module::Metadata)

Additionally a selection of environment variables are also displayed

# Configuration and Environment

Defines the following attributes;

- `env_vars`

    An array reference of environment variable names displayed on the test report.
    Set using the multivalued initialisation argument `env_var`. If the list
    has no values then nothing is printed. An empty list can be set with

        [Test::ReportMetadata]
        env_var = none

- `excludes`

    An array reference of module names to exclude from the test report.
    Set using the multivalued initialisation argument `exclude`

- `includes`

    An array reference of module names to include in the test report.
    Set using the multivalued initialisation argument `include`

- `verify_prereqs`

    A boolean defaulting to true. If true emits lots of warnings if prerequisites
    are not satisfied

# Subroutines/Methods

## `gather_files`

Required by [Dist::Zilla::Role::FileGatherer](https://metacpan.org/pod/Dist::Zilla::Role::FileGatherer)

## `mvp_multivalue_args`

Returns a list of configuration attribute names that are treated as
multi valued

## `register_prereqs`

Required by [Dist::Zilla::Role::PrereqSource](https://metacpan.org/pod/Dist::Zilla::Role::PrereqSource)

# Diagnostics

None

# Dependencies

- [Data::Dumper](https://metacpan.org/pod/Data::Dumper)
- [Data::Section](https://metacpan.org/pod/Data::Section)
- [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla)
- [Moose](https://metacpan.org/pod/Moose)
- [Sub::Exporter::ForMethods](https://metacpan.org/pod/Sub::Exporter::ForMethods)

# Incompatibilities

There are no known incompatibilities in this module

# Bugs and Limitations

There are no known bugs in this module. Please report problems to
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dist-Zilla-Plugin-Test-ReportMetadata.
Patches are welcome

# Acknowledgements

Larry Wall - For the Perl programming language

# Author

Peter Flanigan, `<pjfl@cpan.org>`

# License and Copyright

Copyright (c) 2016 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See [perlartistic](https://metacpan.org/pod/perlartistic)

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE


Changes
Revision history for Dist::Zilla::Plugin::Test::ReportMetadata

          - Toolchain update

0.4.1     2015-10-01 15:54:53
          - Fewer test dependencies in boilerplate
          - Smarter changelog parsing in git hook

0.3.0     2015-09-11 13:28:37
          - Try without Travis helpers
          - Get Travis to install Devel::Cover
          - Added coverage statistics to Travis output
          - Trying Travis CI containers
          - Upstream toolchain update
          - Better prereq data for build and test phases
          - Added broken smoker to exception list

0.2.0     2015-07-16 13:28:21
          - Renamed to ReportMetadata
          - Added project files to master branch



t/boilerplate.pm
package t::boilerplate;

use strict;
use warnings;
use File::Spec::Functions qw( catdir updir );
use FindBin               qw( $Bin );
use lib               catdir( $Bin, updir, 'lib' ), catdir( $Bin, 'lib' );

use Module::Build;
use Sys::Hostname;

sub plan (;@) {
   $_[ 0 ] eq 'skip_all' and print '1..0 # SKIP '.$_[ 1 ]."\n" and exit 0;
}

my ($builder, $host, $notes, $perl_ver);

BEGIN {
   $host     = lc hostname;
   $builder  = eval { Module::Build->current };
   $notes    = $builder ? $builder->notes : {};
   $perl_ver = $notes->{min_perl_version} || 5.008;

   eval { require Test::Requires }; $@ and plan skip_all => 'No Test::Requires';

   $Bin =~ m{ : .+ : }mx and plan skip_all => 'Two colons in $Bin path';

   if ($notes->{testing}) {
      $host eq 'vm-debian-001' and plan skip_all =>
         'Broken smoker 04fd9956-2ca5-11e5-8be5-d039436f2841';
   }
}

use Test::Requires "${perl_ver}";
use Test::Requires { version => 0.88 };

sub import {
   strict->import;
   $] < 5.008 ? warnings->import : warnings->import( NONFATAL => 'all' );
   return;
}

1;

# Local Variables:
# mode: perl
# tab-width: 3
# End:
# vim: expandtab shiftwidth=3:


Build.PL
use strict;
use warnings;

use Module::Build 0.4004;

my $perl_ver = 5.008;
my $name     = 'Dist-Zilla-Plugin-Test-ReportMetadata'; $name =~ s{ - }{::}gmx;
my $prereqs  = {
  'build_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'Data::Section' => '0.200006',
    'Dist::Zilla' => '5.037',
    'Moose' => '2.1405',
    'MooseX::Role::Parameterized' => '1.08',
    'Sub::Exporter::ForMethods' => '0.100051',
    'perl' => '5.008',
    'version' => '0.88'
  },
  'test_requires' => {
    'File::Spec' => '0',
    'Module::Build' => '0.4004',
    'Module::Metadata' => '0',
    'Sys::Hostname' => '0',
    'Test::Requires' => '0.06',
    'version' => '0.88'
  },
};
my $testing  = sub { $ENV{AUTOMATED_TESTING}
                  || $ENV{PERL_CPAN_REPORTER_CONFIG}
                  || $ENV{PERL5_CPANPLUS_IS_VERSION} };
my $whimper  = sub { print {*STDOUT} $_[ 0 ]."\n"; exit 0 };
my $stop     = sub {
   $testing->() or $whimper->( $_[ 0 ] );
   $prereqs->{build_requires} = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires } = { %{ $prereqs->{configure_requires} } };
   $prereqs->{test_requires }->{ 'Test::Requires' } = 0.06;
   delete $prereqs->{requires};
};

# Not interested in reports below min perl ver
$] < $perl_ver and $stop->( "Minimum required Perl version is ${perl_ver}" );

Module::Build->subclass( code => do { local $/; <DATA> } )->new(
   'module_name' => $name,
   'notes'       => { min_perl_version => $perl_ver, testing => $testing->() },
   %{ $prereqs },

)->create_build_script;

exit 0;

__DATA__

sub _added_to_INC { # CVE-2016-1238
  my $self = shift; my %seen;

  for ($self->_default_INC) { $seen{ $_ }++ }

  return grep { not $seen{ $_ }++ } $self->base_dir, @INC;
}
